File: update-java/src/main/java/com/mqd/updatejava/download/DownloadController.java
package com.mqd.updatejava.download;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * 下载状态控制器（内存单例，替换 Kotlin StateFlow）。
 */
public class DownloadController {

    public enum Status { IDLE, DOWNLOADING, FAILED }

    public static class State {
        public Status status = Status.IDLE;
        public String version = "";
        public int progress = 0;
    }

    public interface Listener {
        void onStateChanged(State state);
    }

    private static final State currentState = new State();
    private static final List<Listener> listeners = new CopyOnWriteArrayList<>();

    public static void addListener(Listener listener) {
        if (listener != null) listeners.add(listener);
    }

    public static void removeListener(Listener listener) {
        listeners.remove(listener);
    }

    public static State getCurrentState() {
        return currentState;
    }

    public static void onStart(String version) {
        synchronized (currentState) {
            currentState.status = Status.DOWNLOADING;
            currentState.version = version;
            currentState.progress = 0;
        }
        notifyListeners();
    }

    public static void onProgress(int percent) {
        synchronized (currentState) {
            currentState.progress = Math.max(0, Math.min(100, percent));
        }
        notifyListeners();
    }

    public static void onFinish() {
        synchronized (currentState) {
            currentState.status = Status.IDLE;
            currentState.version = "";
            currentState.progress = 0;
        }
        notifyListeners();
    }

    public static void onFailed(String version) {
        synchronized (currentState) {
            currentState.status = Status.FAILED;
            currentState.version = version;
        }
        notifyListeners();
    }

    public static void reset() {
        synchronized (currentState) {
            currentState.status = Status.IDLE;
            currentState.version = "";
            currentState.progress = 0;
        }
        notifyListeners();
    }

    private static void notifyListeners() {
        State snapshot;
        synchronized (currentState) {
            snapshot = new State();
            snapshot.status = currentState.status;
            snapshot.version = currentState.version;
            snapshot.progress = currentState.progress;
        }
        for (Listener listener : listeners) {
            listener.onStateChanged(snapshot);
        }
    }
}

File: update-lib/src/main/java/com/mqd/updatelib/download/DownloadController.kt
package com.mqd.updatelib.download

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

/**
 * 主进程内存单例，承载下载「进行中」实时状态。
 *
 * 只表达 IDLE / DOWNLOADING / FAILED —— **没有 DONE**：
 * 「已下完」一律由 apk 文件存在 + size 匹配判定，进程重启不丢。
 * 进度只在这里，绝不进 DataStore（避免写放大）。
 *
 * @author quzhuligpt@gmail.com
 * @since 2026/6/19
 */
object DownloadController {

    enum class DownloadStatus { IDLE, DOWNLOADING, FAILED }

    data class DownloadUiState(
        val status: DownloadStatus = DownloadStatus.IDLE,
        val version: String = "",
        val progress: Int = 0
    )

    private val _flow = MutableStateFlow(DownloadUiState())
    val flow: StateFlow<DownloadUiState> = _flow.asStateFlow()

    fun onStart(version: String) {
        _flow.value = DownloadUiState(DownloadStatus.DOWNLOADING, version, 0)
    }

    fun onProgress(percent: Int) {
        _flow.update { it.copy(progress = percent.coerceIn(0, 100)) }
    }

    /** 下载成功结束：回 IDLE，完成态交给文件判据。 */
    fun onFinish() {
        _flow.value = DownloadUiState()
    }

    fun onFailed(version: String) {
        _flow.update { it.copy(status = DownloadStatus.FAILED, version = version) }
    }

    fun reset() {
        _flow.value = DownloadUiState()
    }
}
